lib/workers/branch/reuse.ts

```diff
@@ -86,5 +87,45 @@ export async function shouldReuseExistingBranch(
     logger.debug(`Branch is not mergeable but can't be rebased`);
   }
   logger.debug(`Branch does not need rebasing`);
+
+  // Branches can get in an inconsistent state if "update-lockfile" is used at the same time as other strategies
+  // On the first execution, everything is executed, but if on a second execution the package.json modification is
+  // skipped but the lockfile update is executed, the lockfile will have a different result than if it was executed
+  // along with the changes to the package.json. Thus ending up with an incomplete branch update
+  // This is why we are skipping branch reuse in this case (#10050)
+  const groupedByPackageFile: Record<string, Set<RangeStrategy>> = {};
+  for (const upgrade of config.upgrades) {
+    groupedByPackageFile[upgrade.packageFile] =
+      groupedByPackageFile[upgrade.packageFile] || new Set();
+    groupedByPackageFile[upgrade.packageFile].add(upgrade.rangeStrategy);
+
+    if (
+      groupedByPackageFile[upgrade.packageFile].size > 1 &&
+      groupedByPackageFile[upgrade.packageFile].has('update-lockfile')
+    ) {
+      logger.debug(
+        `Detected multiple rangeStrategies along with update-lockfile`
+      );
+      return {
+        reuseExistingBranch: false,
+        isModified: false,
+      };
+    }
+  }
+
+  // Branches can get in an inconsistent state if postUpdateOptions is used.
+  // package.json updates are run conditionally, but postUpdateOptions are run everytime
+  // On the first execution, everything is executed, but if on a second execution the package.json modification is
+  // skipped but the postUpdateOptions is executed, the lockfile will have a different result than if it was executed
+  // along with the changes to the package.json. Thus ending up with an incomplete branch update
+  // This is why we are skipping branch reuse when postUpdateOptions is used (#10050)
+  if (config.postUpdateOptions?.length > 0) {
```

Comment:
@viceice this might trigger a lot of extra work, don't you think? e.g. anyone with goModTidy or yarnDedupeX will have their branch retried (including package manager) every run

Reply:
🤔 Yes, i think that can happen, so no nice solution.

So we need to workout, why lockfile update is running on wrong `package.json`.

Reply:
I can modify my PR if we can find another way to achieve the same result.
For now /not/ having this option means that on every Renovate run (every hour in my environment) ALL Prs that have @types changes along with other changes get force-pushed every hour. And re-run all their CI

Reply:
Thinking about it, in the end my issue is fixed by the above check (if multiple rangeStrategy are present in a single package.json) maybe this check can be removed altogether. What do you think ?

Reply:
I removed this part and changed the PR's title